src/clp_ffi_js/ir/utils.hpp

```diff
@@ -0,0 +1,163 @@
+#ifndef CLP_FFI_JS_IR_UTILS_HPP
+#define CLP_FFI_JS_IR_UTILS_HPP
+
+#include <algorithm>
+#include <cstddef>
+#include <string>
+#include <string_view>
+#include <type_traits>
+#include <vector>
+
+#include <clp/ErrorCode.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <clp/TraceableException.hpp>
+#include <clp/type_utils.hpp>
+#include <emscripten/em_asm.h>
+#include <emscripten/val.h>
+
+#include <clp_ffi_js/ClpFfiJsException.hpp>
+#include <clp_ffi_js/constants.hpp>
+#include <clp_ffi_js/ir/StreamReader.hpp>
+#include <clp_ffi_js/ir/StructuredIrStreamReader.hpp>
+#include <clp_ffi_js/ir/UnstructuredIrStreamReader.hpp>
+
+namespace clp_ffi_js::ir {
+constexpr std::string_view cEmptyJsonStr{"{}"};
+
+/**
+ * A templated function that implements `StreamReader::filter_log_events` for both
+ * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`. The additional arguments,
+ * which are not part of `filter_log_events`, are private members of the derived
+ * `StreamReader` classes.
+ *
+ * @param log_level_filter
+ * @param log_events
+ * @param[out] filtered_log_event_map A reference to `FilteredLogEventsMap` that stores filtered
+ * result.
+ */
+template <typename LogEvents>
+auto generic_filter_log_events(
+        FilteredLogEventsMap& filtered_log_event_map,
+        LogLevelFilterTsType const& log_level_filter,
+        LogEvents const& log_events
+) -> void {
+    if (log_level_filter.isNull()) {
+        filtered_log_event_map.reset();
+        return;
+    }
+
+    filtered_log_event_map.emplace();
+    auto filter_levels
+            = emscripten::vecFromJSArray<std::underlying_type_t<LogLevel>>(log_level_filter);
+
+    for (size_t log_event_idx = 0; log_event_idx < log_events.size(); ++log_event_idx) {
+        auto const& log_event = log_events[log_event_idx];
+        if (std::ranges::find(
+                    filter_levels,
+                    clp::enum_to_underlying_type(log_event.get_log_level())
+            )
+            != filter_levels.end())
+        {
+            filtered_log_event_map->emplace_back(log_event_idx);
+        }
+    }
+}
+
+/**
+ * A templated function that implements `StreamReader::decode_range` for both
+ * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`. The additional
+ * arguments, which are not part of `decode_range`, are private members of
+ * the derived `StreamReader` classes.
+ *
+ * @param begin_idx
+ * @param end_idx
+ * @param filtered_log_event_map
+ * @param log_events
+ * @param use_filter
+ * @param ts_pattern A pattern for formatting unstructured log event timestamps as date strings.
+ * The pattern is unused for structured log events, as they are not formatted in `clp-ffi-js`.
+ * @return
+ */
+template <typename LogEvents>
+auto generic_decode_range(
```

Comment:
I haven't tried this, so maybe there's some unforeseen limitation, but I think a clean approach is:

* Make StreamReader a template class that takes the log event type as a template parameter.
* Add an abstract method to `StreamReader` for converting a log event into a string---this is where the derived classes can implement non-generic part of `generic_decode_range` will go.
* Make `decode_range` and `filter_log_events` non-abstract methods of `StreamReader`, and use the util function implementations as the method implementations.
* In `decode_range`, call the new method you added.

Reply:
Can we move those into StreamReader? I think the only difference (besides the template parameter) is that one is a shared_ptr and one isn't, right?

Reply:
Right, Marco mentioned that we can only embind an instantiated type, so changing decode_range to use a lambda is probably the cleaner approach.